In [94]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [95]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [96]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [97]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90,
    # Adding descriptors 
    title =  "Crypto Market Data",
    ylabel = "Return (%)",
    xlabel = "Crypto Name",
    value_label= "Return in the Period (%)"
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (Return in the Period (%))

---

### Prepare the Data

In [98]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaled = StandardScaler().fit_transform(df_market_data)

In [99]:
# Create a DataFrame with the scaled data
scaledmarket_df = pd.DataFrame (
    scaled,
    columns= df_market_data.columns
)
# Copy the crypto names from the original data
scaledmarket_df ["coin_id"] = df_market_data.index

# Set the coinid column as index
scaledmarket_df = scaledmarket_df.set_index("coin_id")

# Display sample data
scaledmarket_df.head()


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [100]:
# Create a list with the number of k-values from 1 to 11
k = range(1,12)
list(k)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [101]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for i in k:
    # Inside the loop:
    
    # 1. Create a KMeans model using the loop counter for the n_clusters
    model = KMeans(n_clusters=i)
    # 2. Fit the model to the data using `df_market_data_scaled`
    model.fit(scaledmarket_df)
    # 3. Append the model.inertia_ to the inertia list
    inertia.append(model.inertia_)


c:\Users\gbnlo\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1037: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  "KMeans is known to have a memory leak on Windows "


In [102]:
# Inertia
inertia

[286.99999999999994,
 195.82021818036043,
 123.19048183836955,
 79.02243535120974,
 63.85866780584264,
 52.73890071872584,
 43.67780470293914,
 36.791069782922925,
 32.111296107030306,
 28.229506198258065,
 24.852589613695525]

In [103]:
# Create a dictionary with the data to plot the Elbow curve
dic_inertia = {'k':k, 'inertia':inertia}

# Create a DataFrame with the data to plot the Elbow curve
inertia_df = pd.DataFrame(dic_inertia)

In [104]:
# INERTIA_DF display

inertia_df

k     inertia
0    1  287.000000
1    2  195.820218
2    3  123.190482
3    4   79.022435
4    5   63.858668
5    6   52.738901
6    7   43.677805
7    8   36.791070
8    9   32.111296
9   10   28.229506
10  11   24.852590

In [105]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
inertia_plot = inertia_df.hvplot(
    x = 'k',
    y = 'inertia',
    title = 'Standarized Performance Data: Inertia versus Numbr of Clusters',
    xlabel = 'Number of Clusters (k)',
    ylabel = 'Inertia (units)'
)

In [106]:
# HV Plot
inertia_plot

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:**  

The best value is k=4, based on the elbow rule.

- significant reduction of inertia of 15% from k=3 to k=4
- its also noticable in the hvplot after k=4
- below is the percentage decrease calculation


In [107]:
# To help answer the question I think finding the decrease rate would be helpful
# Finding inertia for one cluster and finding the rate decrease after
inertia_df ['inertia_for_cluster'] = inertia_df['inertia']/inertia_df.iloc[0,1]

In [108]:
inertia_df.head(2) 

k     inertia  inertia_for_cluster
0  1  287.000000               1.0000
1  2  195.820218               0.6823

In [109]:
# Finding decrease rate
inertia_df['decrease_rate'] = inertia_df['inertia_for_cluster'] - inertia_df['inertia_for_cluster'].shift()

In [110]:
inertia_df

k     inertia  inertia_for_cluster  decrease_rate
0    1  287.000000             1.000000            NaN
1    2  195.820218             0.682300      -0.317700
2    3  123.190482             0.429235      -0.253065
3    4   79.022435             0.275339      -0.153896
4    5   63.858668             0.222504      -0.052835
5    6   52.738901             0.183759      -0.038745
6    7   43.677805             0.152187      -0.031572
7    8   36.791070             0.128192      -0.023996
8    9   32.111296             0.111886      -0.016306
9   10   28.229506             0.098361      -0.013525
10  11   24.852590             0.086594      -0.011766

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [111]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4)

In [112]:
# Fit the K-Means model using the scaled data
model.fit(scaledmarket_df)

KMeans(n_clusters=4)

In [113]:
# Predict the clusters to group the cryptocurrencies using the scaled data
clusters = model.predict(scaledmarket_df)
# Print the resulting array of cluster values.
clusters

array([3, 3, 0, 0, 3, 3, 3, 3, 3, 0, 0, 0, 0, 3, 0, 3, 0, 0, 3, 0, 0, 3,
       0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 2, 3, 0, 0, 1, 0, 0, 0, 0])

In [114]:
# Create a copy of the DataFrame
scaled_copy = scaledmarket_df.copy()

In [115]:
# Add a new column to the DataFrame with the predicted clusters
scaled_copy['cluster_original'] = clusters

In [116]:
# Display sample data
scaled_copy.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  cluster_original  
coin_id                                                     
bitcoin                        -0.251637                 3  
ethereum                       -0.199352                 3  
tether                         -0.282061                 0  
ripple                         -0.295546                 0  
bitcoin-cash                   -0.270317                 3

In [117]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
copy_plot = scaled_copy.hvplot.scatter(
    x = "price_change_percentage_24h",
    y = "price_change_percentage_7d",
    by = 'cluster_original',
    title = "Cryptocurrencies Standarized Returns. K-Mean Clusters w/ k=4",
    hover_cols = 'coid_id'
)


In [118]:
# Hv Plot of stuff
copy_plot

:NdOverlay   [cluster_original]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)

---

### Optimize Clusters with Principal Component Analysis.

In [119]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components = 3)

In [142]:
scaledmarket_df.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

In [120]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
market_data_pca = pca.fit_transform(scaledmarket_df)

# View the first five rows of the DataFrame. 
market_data_pca[0:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [121]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

In [122]:
pca.explained_variance_ratio_.sum()

0.8950316570309842

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** Total Variance = 89.5% 

In [123]:
# Create a new DataFrame with the PCA data.
df_market_data_pca = pd.DataFrame(market_data_pca, columns=['PC1', 'PC2', 'PC3'])

# Copy the crypto names from the original data
df_market_data_pca['coin_id'] = scaledmarket_df.index

# Set the coin_id column as index
df_market_data_pca = df_market_data_pca.set_index('coin_id')


In [124]:
df_market_data_pca.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [125]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,12))
k

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [126]:
# Create an empy list to store the inertia values
inertia_pca = []

# Create a for loop to compute the inertia with each possible value of k
for i in k:
    # Inside the loop:
    # 1. Create a KMeans model using the loop counter for the n_clusters
    model_pca = KMeans(n_clusters = i)
    # 2. Fit the model to the data using `df_market_data_pca`
    model_pca.fit(df_market_data_pca)
    # 3. Append the model.inertia_ to the inertia list
    inertia_pca.append(model_pca.inertia_)

c:\Users\gbnlo\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1037: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  "KMeans is known to have a memory leak on Windows "


In [127]:
inertia_pca

[256.8740855678924,
 168.8118963068521,
 93.77462568057297,
 49.66549665179736,
 37.878747033462474,
 27.61897178795743,
 21.134056037473606,
 17.285460350513915,
 13.75581419678007,
 10.559357562793434,
 8.070352545435469]

In [128]:
# Create a dictionary with the data to plot the Elbow curve
inertia_pca = {'k':k, 'inertia_pca':inertia_pca}

# Create a DataFrame with the data to plot the Elbow curve
inertia_pca_df = pd.DataFrame(inertia_pca)

inertia_pca_df = inertia_pca_df.set_index('k')

In [129]:
inertia_pca_df

inertia_pca
k              
1    256.874086
2    168.811896
3     93.774626
4     49.665497
5     37.878747
6     27.618972
7     21.134056
8     17.285460
9     13.755814
10    10.559358
11     8.070353

In [130]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
inertia_pca_plot = inertia_pca_df['inertia_pca'].hvplot(
    title =  'Principal Components of the Standarized Performance: Inertia v/s Number of Clusters',
    xlabel = 'Number of Clusters (k)',
    ylabel = 'Inertia (units)'
)

inertia_pca_plot

:Curve   [k]   (inertia_pca)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** It is also k = 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** The best k value is still 4 however in this instance 4 is at a lower level of inertia

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [131]:
# Initialize the K-Means model using the best value for k
k = 4
model_pca = KMeans(n_clusters=k, random_state=1)


In [132]:
# Fit the K-Means model using the PCA data
model_pca.fit(df_market_data_pca)

KMeans(n_clusters=4, random_state=1)

In [133]:
# Predict the clusters to group the cryptocurrencies using the PCA data
cluster = model_pca.predict(df_market_data_pca)

# Print the resulting array of cluster values.
cluster[0:5]

array([1, 1, 0, 0, 1])

In [134]:
# Create a copy of the DataFrame with the PCA data
df_marketdata_pca_copy = df_market_data_pca.copy()

# Add a new column to the DataFrame with the predicted clusters
df_marketdata_pca_copy['cluster_pca'] = cluster

# Display sample data
df_marketdata_pca_copy.head()

PC1       PC2       PC3  cluster_pca
coin_id                                                
bitcoin      -0.600667  0.842760  0.461595            1
ethereum     -0.458261  0.458466  0.952877            1
tether       -0.433070 -0.168126 -0.641752            0
ripple       -0.471835 -0.222660 -0.479053            0
bitcoin-cash -1.157800  2.041209  1.859715            1

In [143]:
# merging dataframes for plot
result = pd.concat([df_marketdata_pca_copy, scaledmarket_df], axis = 1)

In [144]:
result.describe()

PC1           PC2           PC3  cluster_pca  \
count  4.100000e+01  4.100000e+01  4.100000e+01    41.000000   
mean  -2.301682e-17  9.274424e-17 -2.437075e-17     0.439024   
std    1.633706e+00  1.577905e+00  1.123865e+00     0.672636   
min   -1.157800e+00 -3.896891e+00 -1.986985e+00     0.000000   
25%   -5.931422e-01 -4.944088e-01 -6.633884e-01     0.000000   
50%   -4.502111e-01 -1.510195e-01 -1.149420e-01     0.000000   
75%   -3.062723e-01  4.929762e-01  6.003029e-01     1.000000   
max    8.089018e+00  6.767679e+00  2.846143e+00     3.000000   

       price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                4.100000e+01   
mean                      0.000000                3.520219e-17   
std                       1.012423                1.012423e+00   
min                      -4.981042               -1.682027e+00   
25%                      -0.127467               -7.066688e-01   
50%                       0.077497               -1.906843e-01   
75%                       0.331280                4.931931e-01   
max                       1.919812                2.572251e+00   

       price_change_percentage_14d  price_change_percentage_30d  \
count                 4.100000e+01                 4.100000e+01   
mean                 -8.123583e-17                 5.144936e-17   
std                   1.012423e+00                 1.012423e+00   
min                  -2.217108e+00                -1.393153e+00   
25%                  -6.299628e-01                -4.605580e-01   
50%                  -9.190922e-03                -6.103015e-02   
75%                   6.435649e-01                 1.165382e-01   
max                   2.907054e+00                 5.351455e+00   

       price_change_percentage_60d  price_change_percentage_200d  \
count                 4.100000e+01                  4.100000e+01   
mean                  2.978647e-17                 -1.435166e-16   
std                   1.012423e+00                  1.012423e+00   
min                  -9.560487e-01                 -5.511464e-01   
25%                  -5.517599e-01                 -4.998478e-01   
50%                  -1.592496e-01                 -3.550537e-01   
75%                   1.606038e-02                 -4.736110e-02   
max                   4.769913e+00                  4.632380e+00   

       price_change_percentage_1y  
count                4.100000e+01  
mean                 4.129488e-17  
std                  1.012423e+00  
min                 -2.963296e-01  
25%                 -2.817468e-01  
50%                 -2.255326e-01  
75%                 -1.454693e-01  
max                  6.088625e+00

In [146]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
copy_pca_plot = result.hvplot.scatter(
    x = "price_change_percentage_24h",
    y = "price_change_percentage_7d",
    by = 'cluster_pca',
    hover_cols = 'coid_id'
)


In [147]:
copy_pca_plot

:NdOverlay   [cluster_pca]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [152]:
# Composite plot to contrast the Elbow curves
inertia_plot + inertia_pca_plot

:Layout
   .Curve.I           :Curve   [k]   (inertia)
   .Curve.Inertia_pca :Curve   [k]   (inertia_pca)

In [153]:
# Composite plot to contrast the clusters
copy_plot + copy_pca_plot

:Layout
   .NdOverlay.I  :NdOverlay   [cluster_original]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)
   .NdOverlay.II :NdOverlay   [cluster_pca]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** When we use fewer features to group data using K-Means, it can have a big impact on the results we see. Using fewer features can make it easier to see clear groups in the data. This is because we are focusing only on the most important information, and removing any noisy or irrelevant data.

However, using fewer features can also mean that we lose some important information and our results may not be as accurate. So, it's important to choose the right number of features to use, based on what we want to achieve and what the data tells us